keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,903 @@
+	NodeTopology *v1.NodeSelector
+
+	// Nodes can be used to describe a storage pool that is available
+	// only for certain nodes in the cluster.
+	//
+	// +listType=set
+	// +optional
+	Nodes []string
+}
+
+// CSIStoragePoolStatus contains runtime information about a CSIStoragePool.
+//
+// It is expected to be extended with other
+// attributes which do not depend on the storage class, like health of
+// the pool. Therefore it has the list of
+// `CSIStorageByClass` instances instead of just the capacity
+// and the storage class being in the spec.
+type CSIStoragePoolStatus struct {
+	// Some information, like the actual usable capacity, may
+	// depend on the storage class used for volumes.
+	//
+	// +patchMergeKey=storageClassName
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=storageClassName
+	// +optional
+	Classes []CSIStorageByClass `patchStrategy:"merge" patchMergeKey:"storageClassName" json:"classes,omitempty" protobuf:"bytes,4,opt,name=classes"`
+}
+
+// CSIStorageByClass contains information that applies to one storage
+// pool of a CSI driver when using a certain storage class.
+type CSIStorageByClass struct {
+	// The storage class name matches the name of some actual
+	// `StorageClass`, in which case the information applies when
+	// using that storage class for a volume. There are also two
+	// special names:
+	// - <ephemeral> for storage used by ephemeral inline volumes (which
+	//   don't use a storage class)
+	// - <fallback> for storage that is the same regardless of the storage class;
+	//   it is applicable if there is no other, more specific entry
+	StorageClassName string `json:"storageClassName" protobuf:"bytes,1,name=storageClassName"`
+
+	// Capacity is the size of the largest volume that currently can
+	// be created. This is a best-effort guess and even volumes
+	// of that size might not get created successfully.
+	// +optional
+	Capacity *resource.Quantity `json:"capacity,omitempty" protobuf:"bytes,2,opt,name=capacity"`
+}
+
+const (
+	// FallbackStorageClassName is used for a CSIStorage element which
+	// applies when there isn't a more specific element for the
+	// current storage class or ephemeral volume.
+	FallbackStorageClassName = "<fallback>"
+
+	// EphemeralStorageClassName is used for storage from which
+	// ephemeral volumes are allocated.
+	EphemeralStorageClassName = "<ephemeral>"
+)
+```
+
+This approach has the advantage that the size of the `CSIStoragePool`
+objects do not depend on the number of nodes in the cluster for local
+storage.
+
+The downsides are:
+- Some attributes (driver name, topology) must be stored multiple times
+  compared to a single, more complex object that contains all information
+  for one CSI driver, so overall size in etcd is higher.
+- Higher number of objects which all need to be retrieved by a client
+  which does not already know which `CSIStoragePool` has a certain spec.
+- Constraints like "unique spec" cannot be enforced by the API server.
```

Comment:
Each pool has a unique name?

Reply:
Yes. It has to have a unique name because there can be only one object with a given name. This is enforced by the API server.

What isn't caught at the time when a new pool object is created is when an existing `csistoragepool/a` and the new `csistoragepool/b` have the same spec.